xblock/mixins.py

```diff
@@ -610,3 +611,43 @@ def has_support(self, view, functionality):
             True or False
         """
         return hasattr(view, "_supports") and functionality in view._supports  # pylint: disable=protected-access
+
+
+class CompletableXBlockMixin(object):
+    """
+    This mixin sets attributes and provides helper method to integrate XBlock with Completion API.
+    """
+
+    has_custom_completion = True
+    completion_method = 'completable'
+
+    def emit_completion(self, completion_ratio):
+        """
+        Emits completion event through Completion API.
+
+        Unlike grading API, calling this method allows completion to go down - i.e. emitting a value of 0.0 on
+        a previously completed block indicates that it is no longer considered complete.
+
+        Arguments:
+            completion_ratio (float): Ratio of completion within [0.0; 1.0] range (inclusive). 0.0 means the block
```

Comment:
We agreed to user the term "percent" instead of "ratio".  In the model, this is just called "completion".  I'm option to "completion", "percent_complete", and "completion_percent", but I'd like to be consistent where and if this appears again (if it hasn't already), so maybe @jcdyer can weigh in.  

Note: For consistency, I'd change the doc string to read: "Percent of completion within [0.0; 1.0] range..."

Reply:
@robrap Please forgive me being pedantic, but "percent" is in range [0.0; 100.0], not [0.0; 1.0]. So just renaming the parameter but keeping allowed range would be extremely misleading. Please consider either:

1. Keeping this code intact and renaming other occurences to use mathematically correct term.
2. Change both name and allowed range (if I'm not mistaking it also involves changes down the stream).

Of course I can just rename it to "percent", but it would be a misleading name.

Reply:
@e-kolpakov I'm totally with you on the distinction, and had originally named it "ratio" on my design doc, but we determined that it's more important to be consistent with edx-platform's general terminology, which is to call floats constrained to [0.0, 1.0] "percent".

Reply:
We've discussed this at length in other conversations, and I think we should continue as previously decided.  

A brief summary:
1. Ratio isn't precise either, because a ratio could be any range, and it is the fact that it is a percent that requires the range of 0 to 1.
2. I don't think it is "extremely" misleading.  There are places in our code as well as other code bases where a percent is stored or passed around as 0 to 1, even though it will finally become 0-100 when it is displayed with a percent sign.

Worst case scenario, someone coding with this for the first time doesn't read the docs, thinks it is 0 to 100, and quickly finds out they are wrong when 1% is the max percent they are able to reach.

Reply:
@robrap Is american math or something? :) Ok, I'll do that, but since I have "minority report" I'll add a link to this discussion.

Regarding design doc - I probably haven't reached the part where you guys discuss "ratio" vs. "percent".

Ok, final two thing before I wrap up this discussion:

1. Percent can be any range also (130% ROI, -15% market price drop, etc.).
2. Ratio is the accurate term here - ratio of 1.0 means "things you're comparing are exactly equal", which in our case corresponds to "you've got maximum score possible".
3. The worst case scenario you described contributes to bad developer experience. Basically having ratios called percents anywhere edx-platform causes:
3.1. The need to translate name into actual meaning each time someone touches the code operating on grades/progress/etc. - i.e. taxes productivity a little bit.
3.2. And taxes productivity heavier when the scenario you described happens - which is actually not limited to first use, but might also be caused by forgetting the fact, rushing for whatever reason (i.e. hotfix) and maybe some more.
3.3. Long and pointless discussions like this one :)

So, as a member of community I flag this "percent" vs. "ratio" problem as something that reduces my productivity. It's just a tiny bit, but probably you guys (core developers and architects) might consider this as part of technical debt and address when/if there's opportunity to do so.

And sure, I understand effort & risk vs. outcome is somewhere in range "not a chance", but still.